### PR TITLE
Implement ExecutionTracingProvider for simpler logging

### PR DESCRIPTION
**Pull request by jvmncs**
These commits introduce an ExecutionTracingProvider that traces the evolution of a federated computation through a given executor stack. We've found this to be helpful for both learning and debugging the core TFF framework. After monkey-patching executors with [logging helpers](https://github.com/tf-encrypted/rfcs/blob/master/20190924-tensorflow-federated/logging_helpers.py) for this purpose, we found the recently introduced TracingProvider mechanism to be ideal for integrating these properly into the library, and thought they might be useful to other newcomers. As noted in the documentation, this implementation does not attempt to handle logs written from asynchronous calls, so use with e.g. the ThreadDelegatingExecutor is not recommended.

### CODE REVIEW

**Comment by jvmncs** [comment]
@michaelreneer @roselander, it looks like I can't assign a reviewer myself per GH org permissions, so pinging you both here.

**Comment by michaelreneer** [comment]
@jvmncs thanks. I'll get this assigned. I'll also @cramertj has been working recently to extend the TracingProvider API to work with a list of providers instead of a single provider.

**Comment by cramertj** [comment]
As @michaelreneer mentioned, I'm nearly done with a refactor of the TraceProvider API which should make the code in this PR much simpler and easier, as well as making it easier to support the thread delegating executor. I'd be happy to work with you to update and land this PR once that's in-- I'll comment here once my change has landed (should be either today or early next week). Thanks for your patience!

**Comment by cramertj** [comment]
I've just landed https://github.com/tensorflow/federated/commit/fbfad15be896aa69d8aaa9aa8d4d921032d8fd05. It changes up the `TracingProvider` interface so that there is now just one required method called `span`, and the details of tracking the context (the `yield`ed value) across tasks and threads should be handled for you. The rework also makes it possible to configure multiple tracing providers at once.

**Comment by cramertj** [comment]
Let me know if you have any questions or want help updating this CL-- I'd be happy to work with you to get it patched since I went in and changed the whole API after you'd already put this together :)

**Comment by jvmncs** [comment]
hi @cramertj, I've updated this PR to the new API. I was unsure where I should put the new files, so they ended up in python/core/utils. if you can suggest better placement, I will be happy to move them.

you will notice a large number of dependencies on the python/core/impl/executors subfolder. there is likely a better approach, but would appreciate feedback on the PR as a whole before working to improve that (i.e. where to register classes with this tracing provider)

**Comment by michaelreneer** [comment]
@jvmncs These files should probably be in common_libs along with the other tracing code.

**Comment by jvmncs** [comment]
@michaelreneer it seems like the dependencies I mentioned (e.g. '//tensorflow_federated/python/core/impl/computation_impl.py`) don't have Bazel visibility outside of the `core` subfolder. so placing these files in the common_lib break the build. any suggestions how to proceed?

**Comment by cramertj** [comment]
I don't see your new changes, but I see that you force-pushed twice above-- is it possible that you force-pushed the old version of this change back?

**Comment by michaelreneer** [comment]
@jvmncs I'll comment in the code.

File: tensorflow_federated/python/core/utils/execution_tracing.py [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+from tensorflow_federated.python.core.impl.compiler import intrinsic_defs
+from tensorflow_federated.python.core.impl.executors import caching_executor
+from tensorflow_federated.python.core.impl.executors import composing_executor
+from tensorflow_federated.python.core.impl.executors import eager_tf_executor
+from tensorflow_federated.python.core.impl.executors import federating_executor
+from tensorflow_federated.python.core.impl.executors import reference_resolving_executor
+from tensorflow_federated.python.core.impl.executors import remote_executor
+from tensorflow_federated.python.core.impl.executors import sizing_executor
+from tensorflow_federated.python.core.impl.executors import thread_delegating_executor
+from tensorflow_federated.python.core.impl.executors import transforming_executor
```
Comment (michaelreneer) [inline tensorflow_federated/python/core/utils/execution_tracing.py:25]: These dependencies will require that this tracing utility have a privileged visibility. I think where it is located is probably the appropriate location given the dependencies. But let me ask, is it possible to remove these dependencies? It seems to me like you are simply stripping the module name from all non-primative types? I think we could do this without having an inclusion list of types that we check against. Additionally, we would not have to come back and add more when we add executors.

File: tensorflow_federated/python/core/utils/execution_tracing.py [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+  # Register all types here
+  def _make_formatting_strategies(self):
+    self.register_executors()
+    self.register_executor_values()
+    self.register_computations()
+    self.register_computation_types()
+    self.register_primitives()
+
+  def register_executors(self):
+    register = self._register_formatting_strategy
```
Comment (michaelreneer) [inline tensorflow_federated/python/core/utils/execution_tracing.py:233]: Nit: I would suggest that we not alias the self describing name to something that is less self describing or we choose that the shorter name is self describing and rename the method to that. The alias seems unnecessary.

File: tensorflow_federated/python/core/utils/execution_tracing.py [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+
+class TraceFilterMode(Enum):
+  VALUES = 0
+  INTRINSICS = 1
+  ALL = 2
+
+
+def pattern_union(*patterns):
+  pattern_strs = map(lambda p: p.pattern, patterns)
+  return re.compile("|".join(pattern_strs))
```
Comment (michaelreneer) [inline tensorflow_federated/python/core/utils/execution_tracing.py:36]: Can you use single quotes throughout the change consistently? I'm not sure that we have 100% parity between internal and external lint, but I know this fails internally.

File: tensorflow_federated/python/core/utils/execution_tracing.py [hunk 4]
```diff
@@ -0,0 +1,313 @@
...
+        None,
+}
+
+DEFAULT_FORMAT_STRATEGY = lambda x: "{} {}".format(x, type(x))
+
+
+class ExecutionTracingProvider(tracing.TracingProvider):
+  """Traces a subset of functions appearing in the current execution context.
+
+  NOTE: This TracingProvider does not attempt to correlate log messages across
```
Comment (michaelreneer) [inline tensorflow_federated/python/core/utils/execution_tracing.py:57]: NOTE -> Note, it renders better with our documentation generation

File: .gitignore [hunk 5]
```diff
@@ -4,6 +4,7 @@
 # Python
 *.py[cod]
 __pycache__/
+venv/
```
Comment (michaelreneer) [inline .gitignore:7]: You could break this out into a different PR, probably not worth it, but if you wanted to this change is unrelated and does not require the same level of review.

**Comment by jvmncs** [comment]
>I don't see your new changes, but I see that you force-pushed twice above-- is it possible that you force-pushed the old version of this change back?

Looks like the changes were left in a stash, ended up with a partial commit. Will push the full code shortly

File: tensorflow_federated/python/core/utils/execution_tracing.py (hunk 1 above)
Comment (jvmncs) [reply tensorflow_federated/python/core/utils/execution_tracing.py:25]: That's a good point -- this code grew out of some hastier monkey-patching code, which used required the actual types. I'll remove them.

On the other hand, I'm trying to think through how we would differentiate between, for example, an executor and an executor value without building up inclusion lists. One possibility would be matching on the type name, but that doesn't feel robust. Do you have any alternatives in mind?

Comment (michaelreneer) [reply tensorflow_federated/python/core/utils/execution_tracing.py]: I'll comment inline in the iteration you pushed.

File: tensorflow_federated/python/common_libs/execution_tracing.py [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+  # Register all types here
+  def _make_formatting_strategies(self):
+    self.register_executors()
+    self.register_executor_values()
+    self.register_computations()
+    self.register_computation_types()
+    self.register_primitives()
+
+  # pylint: disable=unnecessary-lambda,missing-function-docstring
+  def register_executors(self):
```
Comment (michaelreneer) [inline tensorflow_federated/python/common_libs/execution_tracing.py:191]: It seems like all format strategies could be combined based on capabilities instead of class. So instead of:

for executor_base.Executor, EagerTensor
`'<{class} @{id}>'`

for executor_value_base.ExecutorValue
`'<{class} @{id} : {type_signature}>'`

for executor_value_base.AnonymousTuple
`'<{class} @{id} : {value}>'`

for computation_types.Type
`'<{class}>'`

for float, int
`'<{value} : {class}>'`

for None
`'-'`

for list, tuple
`'<{element 1}, {element 2}, ...>'`

for ComputationImpl ComputationPB
`'<{class} {context} @{id} : {type_signature}>'`

for IntrinsicDef
`'<{class} {uri} {type_signature} @{id}>'`

you could use: ` '<{class} {context} @{id}: {value}'`, where:

* special case None
* everything has a `class`
* ComputationImpl and ComputationPB have a `context` based on proto
* IntrinsicDef has a `context` based on URI
* everything else has no `context`
* everything as an `id`
* things with type_signature have a `value` of type_signature
* things with a value have a `value` of value
* everything else has no `value`

This make the problem a lot smaller... And makes the representations more consistent.

Maybe we can use duck-typing to get rid of the dependencies? Or maybe drop those differences? Or maybe fork on base classes instead of the individual subclasses? WDYT?

Comment (jvmncs) [reply tensorflow_federated/python/common_libs/execution_tracing.py:191]: This seems useful! I do think the differences matter to make the end result readable, so I'm inclined to choose duck-typing or forking. I think I slightly prefer duck-typing; we can think of these differences as "traits" that are agnostic to the types, which means we can avoid the imports. The only improvement I can imagine would be to let the types themselves handle this, or perhaps to pass in as trace_opts, but both of those alternatives seem invasive for something that's meant to be a simple & lightweight developer tool.

If there are no objections, I'll refactor the PR with all this in mind in the next day or two.

Comment (michaelreneer) [reply tensorflow_federated/python/common_libs/execution_tracing.py:191]: +1 to "let the types themselves handle this explicitly". This would be nice, and I agree this is more invasive.

File: tensorflow_federated/python/common_libs/execution_tracing.py [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+    self._is_traceable_regex = _TRACE_FILTER_PATTERNS[trace_filter_mode]
+
+    # fill in formatting strategies
+    self._default_strategy = default_format_strategy
+    self._formatting_strategies = dict()
+    self._make_formatting_strategies()
+
+    # indentation management
+    # start at -1 so the first call prints at 0
+    self._current_call_level = -1
```
Comment (cramertj) [inline tensorflow_federated/python/common_libs/execution_tracing.py:86]: I was kind of hoping that the indentation level here could be moved into the span yield infra. That is, that the output data could be logically organized underneath its parent by using the parenting abilities offered by `span`. This would mean that either (a) output wouldn't be available to the end of the program or (b) the output would be printed sort of "backwards", listing parent spans *below* their children. Both of these are sort of unfortunate, but either would make this provider compatible with the more complicated (but more common) executor setups involving concurrency. What do you think about this idea? I feel somewhat uneasy about offering a debugging utility that will print incorrect results for most users' default configurations.

File: tensorflow_federated/python/common_libs/execution_tracing.py [hunk 8]
```diff
@@ -0,0 +1,266 @@
...
+      # Pre-fn
+      self._current_call_level += 1
+      self._log_trace('call', scope, sub_scope, fn_args, fn_kwargs)
+      yield None
+
+      # Post-fn
+      self._current_call_level -= 1
+      self._log_trace('retr', scope, sub_scope, fn_args, fn_kwargs)
+
+  @classmethod
```
Comment (cramertj) [inline tensorflow_federated/python/common_libs/execution_tracing.py:117]: Nit: since this doesn't use the class, I don't think this belongs as a class method. It seems more like a private helper function, such as `_pattern_union` above.

File: tensorflow_federated/python/common_libs/execution_tracing.py (hunk 7 above)
Comment (jvmncs) [reply tensorflow_federated/python/common_libs/execution_tracing.py:86]: I like this idea a lot, I hadn't quite connected those dots before. When I get a chance to revisit this PR, I'll give this a shot

**Comment by roselander** [comment]
@cramertj @michaelreneer Can you remove me from this PR? I was _very_ confused when I looked at my GH dashboard today :)

**Review by roselander — APPROVED** [review]

**Comment by cramertj** [comment]
Closing due to inactivity, but feel free to re-open!
